src/storage/structs.cairo

```diff
@@ -1,23 +1,66 @@
-use core::num::traits::Zero;
+use core::num::traits::{Pow, Zero};
 use core::poseidon::poseidon_hash_span;
+use core::starknet::storage_access::StorePacking;
 use onchain_id_starknet::storage::storage::StorageArrayFelt252;
 use starknet::ContractAddress;
-use starknet::storage::{Mutable, StoragePath, StoragePointerWriteAccess};
 use starknet::storage_access::{
     StorageBaseAddress, storage_address_from_base, storage_base_address_from_felt252,
 };
 use starknet::{Store, SyscallResult, SyscallResultTrait};
 
-// TODO: Go over comments
-#[starknet::storage_node]
-pub struct Key {
-    /// Array of the key purposes, like 1 = MANAGEMENT, 2 = EXECUTION.
-    pub purposes: StorageArrayFelt252,
-    /// The type of key used, which would be a uint256 for different key types. e.g. 1 = ECDSA, 2 =
-    /// RSA, etc.
-    pub key_type: felt252,
-    /// Hash of the public key or ContractAddress
-    pub key: felt252,
+/// Struct that holds details about key.
+#[derive(Drop, Copy)]
+pub struct KeyDetails {
+    /// 128 bit bitmap. Capable of holding 128 purposes for a single key.
+    pub purposes: u128,
+    /// Indicates the type of the key.
+    pub key_type: u64,
+}
+
+pub impl KeyDetailsPacking of StorePacking<KeyDetails, felt252> {
+    fn pack(value: KeyDetails) -> felt252 {
+        u256 { low: value.purposes, high: value.key_type.into() }.try_into().unwrap()
+    }
+
+    fn unpack(value: felt252) -> KeyDetails {
+        let value_u256: u256 = value.into();
+        KeyDetails { purposes: value_u256.low, key_type: value_u256.high.try_into().unwrap() }
+    }
+}
+
+pub trait BitmapTrait<T> {
+    fn set(bitmap: T, index: usize) -> T;
+    fn unset(bitmap: T, index: usize) -> T;
+    fn get(bitmap: T, index: usize) -> bool;
+}
+
+impl BitmapTraitImpl of BitmapTrait<u128> {
+    fn set(bitmap: u128, index: usize) -> u128 {
+        bitmap | 2_u128.pow(index)
+    }
+
+    fn unset(bitmap: u128, index: usize) -> u128 {
+        bitmap & (~2_u128.pow(index))
+    }
+
+    fn get(bitmap: u128, index: usize) -> bool {
+        (bitmap & 2_u128.pow(index)).is_non_zero()
+    }
+}
```

Comment:
Consider implementing these as methods that takes a self reference, and implement `Default` or new to create an instance.